config: merge settings and add a .NET config tool

lahm86 opened the pull request:
Part of #633.

#### Checklist

- [x] I have read the [coding conventions](https://github.com/rr-/Tomb1Main/blob/master/CONTRIBUTING.md#coding-conventions)
- [x] I have added a changelog entry about what my pull request accomplishes, or it is an internal change

#### Description

The config tool allows editing `Tomb1Main.json5` and internal changes mean that all settings that are set inside the game are also now stored in this file. The main changes to the game to support this move are detailed below.

* All settings previously loaded/saved in `settings.c` have been moved to `config.c`.
* `Config_Init` function has been added to set specific settings at the same stage as they were being set in `Settings_Read`. This is primarily `GFX_Context_SetVSync`, as calling this during `Config_Read` causes a crash because `Output_Init` has not been called at that stage.
* Some `bool` settings have been inverted in order to maintain readability and consistency in the config tool GUI.
* `Config_Write` replaces `Settings_Write` throughout, and this writes *all* global settings to file.

The comments in the default config file have been updated in line with the GUI; this may be redundant, but it maintains consistency for those who may not wish to use the config tool.

lahm86 committed:
config: merge settings into main config

config:
 - settings.c functionality moved to config.c;
 - some bool settings inverted in line with config tool;
 - default config file comments updated in line with config tool.

Part of #633.

lahm86 committed:
config: add a .NET config tool

Part of #633.

rr- commented on bin/cfg/Tomb1Main.json5:
```diff
@@ -220,19 +220,19 @@
     "enable_enhanced_saves": true,
 
-    // Sound effects are randomly, slightly pitched to vary the game sounds.
+    // Allows sound effects to be randomly, slightly pitched to vary the game sounds.
```
Do we need the comments in this file considering that it's designed to get written back to, and all the comments will get discarded?

Actually, do we even need to ship this file, considering that 
- the user is expected to use the config tool
- the game should do just fine without it, using the defaults for each setting
- the game will create the config file for the power users upon first launch

?

lahm86 replied:
I'd be in favour of removing the file altogether - it'd be one less place to have to define future settings. The only issue I can think of is if you install the game and launch the config tool before running the game - in this case it won't generate the file by default, you would have to open a file manually. But this could easily be changed so that it does generate it. Or the expectation could be that the user must launch the game once before running the config tool?

rr- replied:
I'd rather have the config tool work ootb without requiring the user to launch the game (and use good defaults).

lahm86 replied:
The default configuration file is now removed. The config tool will launch normally if it can't locate the default file, and it will have all properties initialised to their defaults (as defined in `specification.json` in the project resources). If you make a change in the tool and save, it will create the file. See [b27bc7c](https://github.com/rr-/Tomb1Main/pull/637/commits/b27bc7c9fb0f28786dbbb609a64194a4238523ff).

I had to make a small change in the game as without the default config file it wouldn't load. I believe the issue was that the screen resolution wasn't being set. When it couldn't find the config file, it was jumping to the `cleanup` label, so while the default settings were being loaded, it was skipping over the `g_AvailableResolutions` initialisation.
See [4ee0f5a](https://github.com/rr-/Tomb1Main/pull/637/commits/4ee0f5a3487067af292d756b533979a5b9ffd017).

I adjusted the log message here to become a warning rather than an error, and it will now pass `"{}"` to `Config_ReadFromJSON` instead of an empty string to avoid a further error message in that function. The warning message line breaches the 80 column rule so `make lint` shifted it over 3 lines. The wording can perhaps be amended here if this is not ideal.

rr- commented on CHANGELOG.md:
```diff
@@ -1,6 +1,7 @@
 ## [Unreleased](https://github.com/rr-/Tomb1Main/compare/stable...develop)
 - added graphics effects, lava emitters, flame emitters, and waterfalls to the savegame so they now persist on load (#418)
 - added an option to turn off sound effect pitching (#625)
+- added a .NET-based configuration tool (#633)
```
I'd move it to the top, as it's the crown of this release.

lahm86 replied:
Resolved in [b27bc7c](https://github.com/rr-/Tomb1Main/pull/637/commits/b27bc7c9fb0f28786dbbb609a64194a4238523ff)

rr- commented on src/config.c:
```diff
@@ -230,3 +303,115 @@ bool Config_Read(void)
+    Memory_FreePointer(&data);
+
+    return true;
+}
```
Missing end of the line character

lahm86 replied:
Resolved in [4ee0f5a](https://github.com/rr-/Tomb1Main/pull/637/commits/4ee0f5a3487067af292d756b533979a5b9ffd017)

rr- commented on src/config.h:
```diff
@@ -130,3 +130,5 @@ extern CONFIG g_Config;
 bool Config_ReadFromJSON(const char *json);
 bool Config_Read(void);
+void Config_Init(void);
+bool Config_Write(void);
```
Missing end of the line character

lahm86 replied:
Resolved in [4ee0f5a](https://github.com/rr-/Tomb1Main/pull/637/commits/4ee0f5a3487067af292d756b533979a5b9ffd017)

rr- commented on tools/config/Tomb1Main_ConfigTool/Models/Specification/Specification.cs:
```diff
@@ -0,0 +1,44 @@
+    public Specification(string sourceData)
+    {
+        JObject data = JObject.Parse(sourceData);
+        Dictionary<string, List<string>> enumTypes = JsonConvert.DeserializeObject<Dictionary<string, List<string>>>(data[nameof(Enums)].ToString());
```
The lines seem to be a bit long, can you run it through some autoformatter? (Optional, best C# autoformatters are proprietary AFAIR)

lahm86 replied:
Updated in [87bf6ae](https://github.com/rr-/Tomb1Main/pull/637/commits/87bf6ae5de281eb53de6382941738394d8b1b2f6).
This should be a bit more readable now. Unfortunately the second `JsonConvert.DeserializeObject` call has to remain as it requires the custom property converter, whereas the enums can be parsed more simply with `JObject`.

lahm86 committed:
config: remove default config file

config: default configuration file is no longer required; config will load defaults automatically.

lahm86 committed:
config: remove default config file

lahm86 committed:
config: remove default config file

lahm86 committed:
config: improve config tool spec loading

rr- reviewed (approved):
LGTM. Now we need to integrate with GH Actions. My suggestion is to merge this PR and have me create a follow-up PR that integrates the release GH Action. What do you think?

lahm86 commented:
> LGTM. Now we need to integrate with GH Actions. My suggestion is to merge this PR and have me create a follow-up PR that integrates the release GH Action. What do you think?

That sounds good to me, thanks!
